src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -55,19 +56,78 @@ internal static bool TryCombineTypeResolutionsForTarget(QsQualifiedName target,
-        /// the type parameters found.
-        ///
-        /// The given resolutions are expected to be ordered such that dictionaries containing type parameters that take a
-        /// dependency on other type parameters in other dictionaries appear before those dictionaries they depend on.
-        /// I.e., dictionary A depends on dictionary B, so A should come before B. When using this method to resolve
-        /// the resolutions of a nested expression, this means that the innermost resolutions should come first, followed by
-        /// the next innermost, and so on until the outermost expression is given last.
-        ///
-        /// Returns the constructed dictionary as out parameter. Returns true if the combination of the given resolutions is valid,
-        /// i.e. if there are no conflicting resolutions and type parameters are uniquely resolved to either a concrete type, a
-        /// type parameter belonging to a different callable, or themselves.
+        /// Applies all the type parameter resolutions in a dictionary to the whole dictionary,
+        /// resulting in a resolution dictionary that has type parameter keys that are not
+        /// referenced in the values of the dictionary.
         /// </summary>
+        internal static bool TryCombineTypeResolutionDictionary(out TypeParameterResolutions combined, TypeParameterResolutions resolutionDictionary)
+        {
+            if (!resolutionDictionary.Any())
+            {
+                combined = TypeParameterResolutions.Empty;
+                return true;
+            }
+
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static bool IsConstrictiveResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                ConstrictionCheck.Apply(typeParam, res);
+
+            foreach (var (typeParam, paramRes) in resolutionDictionary)
+            {
+                // ToDo: make this using LINQ
+                var temp2 = new List<KeyValuePair<Tuple<QsQualifiedName, NonNullable<string>>, Tuple<QsQualifiedName, NonNullable<string>>>>();
+                foreach (var kvp in combinedBuilder)
+                {
+                    if (ContainsTypeParameter.Apply(kvp.Value))
+                    {
+                        foreach (var temp3 in GetTypeParameters.Apply(kvp.Value))
+                        {
+                            temp2.Add(new KeyValuePair<Tuple<QsQualifiedName, NonNullable<string>>, Tuple<QsQualifiedName, NonNullable<string>>>(kvp.Key, temp3));
+                        }
+                    }
+                }
+                var mayBeReplaced = temp2
+                    .ToLookup(
+                        kvp => kvp.Value,
+                        kvp => kvp.Key);
```

Comment:
I may need some help untangling this and making it a clean LINQ statement.

Reply:
Basically, I have a dictionary that I want to first filter with the `ContainsTypeParameter.Apply`, then each of the remaining KeyValuePairs needs to be mapped through the `GetTypeParameters.Apply` to produce 1 to many new KeyValuePairs per input KeyValuePair. Then I need to take all the resulting KeyValuePairs and create a lookup that swaps keys for values in the pairs.

Reply:
What would be even better is if I could just use `GetTypeParameter.Apply` as the filter, because it returns an empty set if there are no type parameters. Then I could get rid of the `ContainsTypeParamter` transformation which is almost a duplicate of the `GetTypeParameter` transformation. Yes, that would be good.

Reply:
This has been done.